### PR TITLE
[wolverine] Integrate openai into my editor

### PR DESCRIPTION
**Pull request by yacineMTB**
Just goofing around. Going to work with @biobootloader to see if we can get this merged & extend with wolverine.

**I didn't have time to integrate VSCode's node runtime with a keychain. Do NOT add your keys to a json that lives raw on your filesystem**. _Unless_ you have a limit, and you don't care about potentially losing 20 bucks.

Want to pull and build? 
1. pull
2. cd vscode_extension/wolverine
3. npm install
4. npm run package && npm run vsce
4. ctrl shift p => install extensions => the three little dots on top right
5. slurp up the vsix file you just generated

This branch should have a vsix file that you can just use!
If you want to develop, open up vscode in the extension's directory and hit F5

## Demo
https://user-images.githubusercontent.com/10282244/233265760-05af619e-a536-47e9-a061-142261a4167d.mp4

 
# Please don't share on twitter dot com bird app site yet, I'm making a goofy video

### CODE REVIEW

**Comment by yacineMTB** [comment]
https://user-images.githubusercontent.com/10282244/233284504-798258ed-922b-4bc2-b75c-f271ba56615b.mp4

probably overkill for this repo, but computers should probably embed the whole world before doing things, including voice

File: vscode_extension/wolverine/src/extension.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+			let { newCharacterLocation, newLineLocation } = getNewCursorLocation(word, currentLine, currentCharacter);
+			const position = new Position(currentLine, currentCharacter);
+			await activeEditor.edit((editBuilder) => {
+				editBuilder.insert(position, word);
+			});
+			currentCharacter = newCharacterLocation;
+			currentLine = newLineLocation;
+		}
+		// TODO I should make this buffer flush configurable.
+		await sleep(30);
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:92]: i should flush every word so far, latency measured is about 7ms per token

File: vscode_extension/wolverine/src/extension.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+const constructPrompt = async (text: string): Promise<string> => {
+	const defaultPrompt = `
+INSTRUCTIONS: 
+The following text provided has come straight from my text editor. I have an extension that can send my text output to you, often.
+I've added my instructions in the comments, please attend to the comments, and output the code.
+When you output the code, please don't add any human english language. Your entire output must not cause any errors!
+Output code that effectively does the same thing of what I sent you, my extension replaces the text that I've highlighted and sent.
+Remember, you'll have to send me any messages in comments!
+CODE:
+`;
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:105]: I should just go raw from config and figure out how to do defaults with vscode

File: vscode_extension/wolverine/src/extension.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	const configuredPrompt = await workspace.getConfiguration().get('wolverine.prompt');
+	return (configuredPrompt || defaultPrompt) + text;
+};
+
+export async function activate(context: ExtensionContext) {
+
+	let disposable = commands.registerCommand('wolverine.directedHeal', async () => {
+		// Forgive me father..
+		openaikey = await workspace.getConfiguration().get('wolverine.UNSAFE.OpenaiApiKeySetting') || '';
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:114]: there is literally no good solution for this so i will have to roll my own

Comment (yacineMTB) [reply vscode_extension/wolverine/src/extension.ts:114]: i can't believe they don't provide oauth

File: vscode_extension/wolverine/src/extension.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			return;
+		}
+
+		const selection = activeEditor.selection;
+		let range: Range;
+		if (!selection.isEmpty) {
+			range = new Range(selection.start, selection.end);
+		} else {
+			range = activeEditor.document.validateRange(new Range(0, 0, Number.MAX_VALUE, Number.MAX_VALUE));
+		}
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:127]: this is kind of annoying, it should just be a different hotkey for all

File: vscode_extension/wolverine/src/extension.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+const deleteRange = async (activeEditor: TextEditor, range: Range) => {
+	await activeEditor.edit(editBuilder => {
+		editBuilder.delete(range);
+	});
+};
+
+// Yacine threw most of the complexity into here.
+// Holds a buffer in a javascript array, registers a event listener on a server-sent events function, builds the buffer
+// Takes a position, and flushes the buffer on a preconfigured cron into the provided cursor position.
+const useBufferToUpdateTextContentsWhileStreamingOpenAIResponse = async (activeEditor: TextEditor, position: Position, prompt: string) => {
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:67]: I need to take an interrupt into here

Comment (yacineMTB) [reply vscode_extension/wolverine/src/extension.ts:67]: more mutation

File: vscode_extension/wolverine/src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		['Authorization']: `Bearer ${openaikey}`,
+	};
+
+	const configurationModel =  await workspace.getConfiguration().get('wolverine.model');
+	const data = {
+		'model': configurationModel || 'gpt-3.5-turbo',
+		'messages': messages,
+		'temperature': 0.9,
+		'stream': true,
+	};
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:33]: i need to make this configurable

File: vscode_extension/wolverine/src/extension.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	return new Promise(async (resolve) => {
+		const response = await axios({
+			method: 'post',
+			url: 'https://api.openai.com/v1/chat/completions',
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:37]: needs config

File: vscode_extension/wolverine/src/extension.ts (hunk 6 above)
Comment (yacineMTB) [reply vscode_extension/wolverine/src/extension.ts:33]: i should also interface it

File: vscode_extension/wolverine/src/extension.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+			headers: headers,
+			data: data,
+			responseType: 'stream',
+		});
+		response.data.on('data', onDataFunction);
+		response.data.on('end', () => {
+			resolve();
+		});
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:45]: I need to default register a reject and hit a message to the user

File: vscode_extension/wolverine/src/test/runTest.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+		// Download VS Code, unzip it and run the integration test
+		await runTests({ extensionDevelopmentPath, extensionTestsPath });
+	} catch (err) {
+		console.error('Failed to run tests', err);
+		process.exit(1);
+	}
+}
+
+main();
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/test/runTest.ts:23]: bloat

File: vscode_extension/wolverine/src/extension.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+			currentCharacter = newCharacterLocation;
+			currentLine = newLineLocation;
+		}
+		// TODO I should make this buffer flush configurable.
+		await sleep(30); 
+	}
+};
+
+const constructPrompt = async (text: string): Promise<string> => {
+	const defaultPrompt = `
```
Comment (yacineMTB) [inline vscode_extension/wolverine/src/extension.ts:97]: consider taking filename into arg
make configurable

File: vscode_extension/wolverine/package.json [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    "test-compile": "tsc -p ./",
+    "watch": "webpack --watch",
+    "package": "webpack --mode production --devtool hidden-source-map",
+    "compile-tests": "tsc -p . --outDir out",
+    "watch-tests": "tsc -p . -w --outDir out",
+    "pretest": "npm run compile-tests && npm run compile && npm run lint",
+    "lint": "eslint src --ext ts",
+    "test": "node ./out/test/runTest.js"
+  },
+  "devDependencies": {
```
Comment (yacineMTB) [inline vscode_extension/wolverine/package.json:56]: oops i forgot to push new difi

File: vscode_extension/wolverine/src/extension.ts (hunk 3 above)
Comment (nervousapps) [reply vscode_extension/wolverine/src/extension.ts:116]: Hi, in this extension for example [codeGPT](https://github.com/davila7/code-gpt-docs), you can give the key with a user prompt. I don't know how it is stored but it is not hardcoded in a configfile :)

Comment (yacineMTB) [reply vscode_extension/wolverine/src/extension.ts:116]: Thank you! Do you know if the extension processes are sandboxed away from each other?

Honestly, having to copy the buffer in and out probably is same amount of security even if it _was_ sandboxed; given most users will just paste it into a text file for accessibility :P

The real answer here is either keychain (a la copilot) or oauth
The fact that the keys don't expire is the first problem

**Comment by biobootloader** [comment]
Incredible!

Let me get to some of the older PRs before tackling this beast 🔥

**Comment by yacineMTB** [comment]
yeah big guy i'm gonna actually toy around with this and make it not hot garbage
only attend once it's not in draft :) shared it early for someone else who asked for it

**Comment by yacineMTB** [comment]
I was programming something else and was getting a little annoyed about how it didn't have any context. So I made it do what I want!

image

Now if I cast "directed heal" on this selected text, it'll have the references to all visible text. However; it won't include the actual internals of the function. It'll be replaced with ... instead.

Also; here's the new prompt.

```
const constructPrompt = async (text: string, filepath: string): Promise<string> => {
	const defaultPrompt = `
ADDITIONAL CONTEXT:
filepath: ${filepath}
INSTRUCTIONS: 
- The text under 'CODE' has come straight from my text editor. 
- Your entire output MUST be valid code.
- You may communicate back, but they MUST be in comments
- The code sent to you might have some instructions under comments. Follow the instructions. Only attend to instructions contained between [[SELECTED]] & [[/SELECTED]]
- The code between [[SELECTED]] [[/SELECTED]] is code that I need you to replace. So when you start responding, only respond with code that could replace it
- MAKE SURE YOU ONLY WRITE CODE BETWEEN [[SELECTED]] AND [[/SELECTED]]. The rest of it doesnt need to be replace
- The rest of the code is provided as context. Sometimes, I'll collapse my code for you, and hide the details under '...'. Follow the style of the rest of the provided code.
- DO NOT WRITE THE ENTIRE CODE FILE. ONLY REPLACE WHAT IS NECESSARY IN SELECTED
- Prefer functional programming principles.
CODE:
${text}
NEW CODE TO REPLACE WHAT IS BETWEEN [[SELECTED]] and [/SELECTED]:
`;
	const configuredPrompt = await workspace.getConfiguration().get('wolverine.prompt');
	if (configuredPrompt) {
		return configuredPrompt + text;
	}
	return defaultPrompt;
};
```

File: scratch/index.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+const countCharacters = (text) => text.replace(/\n/g, '').length;
+const countNewLines = (text) => text.match(/\n/g)?.length || 0;
+const countTextStream = (textStream) => {
+  const numberOfCharacters = countCharacters(textStream);
+  const numberofNewLines = countNewLines(textStream);
+  return { numberOfCharacters, numberofNewLines };
+};
+
+const main = () => {
+  const textStream = "A B C D E F G H I J K L M N O P Q \n\n R S T U V";
```
Comment (yacineMTB) [inline scratch/index.js:10]: Whoops

**Comment by yacineMTB** [comment]
cc @biobootloader 
FYI I've been using this on my own dev
I'm keeping this branch updated with whatever my patches are as I figure out what works for me vs what doesn't

The code is an awful mess, I plan to encapsulate all of the state transitions.

**Pull request closed without merging**